src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/Pipeline/CaptureLimitPolicyProbes.cs

```diff
@@ -0,0 +1,49 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Diagnostics.Monitoring.HostingStartup.ParameterCapturing.FunctionProbes;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring.HostingStartup.ParameterCapturing.Pipeline
+{
+    internal sealed class CaptureLimitPolicyProbes : IFunctionProbes
+    {
+        private readonly IFunctionProbes _probes;
+        private readonly int _captureLimit;
+        private readonly TaskCompletionSource _stopRequest;
+
+        private int _captureCount;
+        private bool _stopped;
+
+        public CaptureLimitPolicyProbes(IFunctionProbes probes, int captureLimit, TaskCompletionSource stopRequest)
+        {
+            _probes = probes;
+            _captureLimit = captureLimit;
+            _stopRequest = stopRequest;
+        }
+
+        public void CacheMethods(IList<MethodInfo> methods) => _probes.CacheMethods(methods);
+
+        public bool EnterProbe(ulong uniquifier, object[] args)
+        {
+            // In addition to the stop request, use a flag to more quickly react to the limit being reached,
+            // limiting the amount of extra data being captured which is important in the case of hot paths.
+            if (_stopped)
```

Comment:
I'm trying to think if there's a way we can get the probe invocation and capture count exactly right (and not invoke/count more or less), but I'm fine with this as-is; it's better to over-collect than to under-collect.

Reply:
There's a couple ways of ensuring over-collecting doesn't happen, but it's important to note that above all else we want to prioritize minimizing perf impact here, so some minor over-collecting is an acceptable tradeoff for now IMO.

There's likely some larger refactoring we could do to avoid over-collecting while still preserving our perf (e.g. performing the capture limit enforcement in `ParameterCapturingLogger`). I'd prefer to keep things simple for now though until we need to better solve this issue. In the interim if occasional over-collection is a problem, the consumer of this data could always drop the extra messages.